port to PC

brand17 opened the pull request:
Make the library working on PC

tomstewart89 commented:
Hey @brand17, thanks for the PR! 

I actually have some code already that can be used to compile this library on a PC. You can find it here: [Arduino.h](https://github.com/tomstewart89/BasicLinearAlgebra/blob/master/test/Arduino.h). The idea is to set up the inclue directories such that `BasicLinearAlgebra` includes this file rather than Arduino's header file. That header then the `Print` and a few other things. This is how I run the unit tests for this repo (i.e [these guys](https://github.com/tomstewart89/BasicLinearAlgebra/tree/master/test)).

I'd rather stick with that system since it doesn't require any changes to the main library but I notice that it might not perfectly suit your use-case since it prints Serial output to a stringstream rather than stdout. Would you be interested in updating this PR to improve [Arduino.h](https://github.com/tomstewart89/BasicLinearAlgebra/blob/master/test/Arduino.h) instead?

brand17 commented:
Hi, @tomstewart89. Its great idea. I have made a few changes to `Arduino.h` : added `println` and `endl`. I ensured that my changes pass your tests. I am getting error `namespace has no member max`, so I added `#include <algorithm>`. I compiled with g++ on VSCode.

tomstewart89 commented on test/Arduino.h:
```diff
@@ -13,12 +14,23 @@ struct Print
         buf.str("");
     }
 
+    Print& operator<< (std::ostream& (*pf)(std::ostream&)){
+        buf << pf;
```
```suggestion
    Print& operator<< (std::ostream& (*pf)(std::ostream&))
    {
        buf << pf;
```
Sorry to be a stickler, but could we move that open bracket to a new line? Also just to confirm, I guess the idea with this function is to allow a caller to pass things like `std::endl` to `Print` right?

tomstewart89 reviewed (changes requested):
Sorry for the slow review, I just had one pedantic change otherwise this looks good to me! Thanks for the PR!

brand17 commented:
@tomstewart89 Sorry, forgot to answer that the overloaded operator needs to use `<< endl`. You are correct

tomstewart89 commented:
Hey @brand17, did you mean to close this PR?  I’m happy to merge it if you are

brand17 commented:
Hi, @tomstewart89. Sorry - it was my first PR. I thought that I need to close to merge.

tomstewart89 reviewed (approved):
Looks good, thanks for the PR!
